3 query data content

rdahlman3 opened the pull request:
Content for the section on querying data.

Issue #3

Commit:
markdown file started for the querying tutorial

Commit:
Added first section of querying markdown

Commit:
add section for grabbing a whole table

rdahlman3 committed:
section for loading a single entity

rdahlman3 committed:
add section for filtering

rdahlman3 committed:
updates the code snippets to fit our blog/author decision

rdahlman3 committed:
Add table of contents and move querying to the correct directory

helmoski commented on content/efcore/getting-started/querying.md:
```diff
@@ -0,0 +1,47 @@
+# Querying Data
+
+Sometimes you really need an item (or items) from your database based off a certain set of parameters. In case you've forgotten, this is what querying is.
```
This seems patronizing.

helmoski replied:
We should still describe what querying is but in a less patronizing way.

helmoski commented on content/efcore/getting-started/querying.md:
```diff
@@ -0,0 +1,47 @@
+
+Sometimes you really need an item (or items) from your database based off a certain set of parameters. In case you've forgotten, this is what querying is.
+
+Querying with EF Core uses LINQ statements (if you need some more pointers, check [this](https://www.microsoft.com/net/tutorials/csharp/getting-started/linq) out). For those who really like LINQ, [101 LINQ Samples](https://code.msdn.microsoft.com/101-LINQ-Samples-3fb9811b) is a great resource.
```
~~"Querying with EF Core uses LINQ statements (if you need some more pointers, check [this](https://www.microsoft.com/net/tutorials/csharp/getting-started/linq) out)"~~ 
"EF Core uses LINQ to query data. If you need a refresher on LINQ, check out the [LINQ lesson](https://www.microsoft.com/net/tutorials/csharp/getting-started/linq) in the C# Interactive Tutorial."

helmoski commented on content/efcore/getting-started/querying.md:
```diff
@@ -0,0 +1,47 @@
+
+Sometimes you really need an item (or items) from your database based off a certain set of parameters. In case you've forgotten, this is what querying is.
+
+Querying with EF Core uses LINQ statements (if you need some more pointers, check [this](https://www.microsoft.com/net/tutorials/csharp/getting-started/linq) out). For those who really like LINQ, [101 LINQ Samples](https://code.msdn.microsoft.com/101-LINQ-Samples-3fb9811b) is a great resource.
```
"For those who really like LINQ..."
I don't think this is the best reason for why they would want to use this resource. Also, this resource would probably better fit in the LINQ tutorial than this one.

rdahlman3 replied:
Yeah, redirecting to the LINQ tutorial is probably enough for this tutorial.

helmoski commented on content/efcore/getting-started/querying.md:
```diff
@@ -0,0 +1,47 @@
+
+Querying with EF Core uses LINQ statements (if you need some more pointers, check [this](https://www.microsoft.com/net/tutorials/csharp/getting-started/linq) out). For those who really like LINQ, [101 LINQ Samples](https://code.msdn.microsoft.com/101-LINQ-Samples-3fb9811b) is a great resource.
+
+The basic format of queries is:
```
~~queries~~ => a query

helmoski commented on content/efcore/getting-started/querying.md:
````diff
@@ -0,0 +1,47 @@
+The basic format of queries is:
+
+```c#
+variableToStoreQueryResult = context.Model.QueryKeyWord(m => m.Identifier == value)
````
This example might be more confusing than helpful.

helmoski commented on content/efcore/getting-started/querying.md:
````diff
@@ -0,0 +1,47 @@
+The basic format of queries is:
+
+```c#
+variableToStoreQueryResult = context.Model.QueryKeyWord(m => m.Identifier == value)
````
Also, in this context, keyword is one word.

rdahlman3 replied:
I'd like to hear @arynhuck's thoughts on this, since it was originally your writing

helmoski replied:
https://msdn.microsoft.com/en-us/library/x53a06bb.aspx

helmoski commented on content/efcore/getting-started/querying.md:
````diff
@@ -0,0 +1,47 @@
+
+Grabs the entirety of a table.
+
+```c#
````
~~Runnable code should have `{.REPL}` instead of `{c#}`~~

Look at the first code example here: https://raw.githubusercontent.com/LadyNaggaga/training-tutorials/master/content/csharp/getting-started/types-variables.md

We should have both `{.REPL}` and `{.snippet}` code blocks. The snippet code block is what is shown in-text. The REPL code block is what shows up in the code editor.

rdahlman3 replied:
At this time, we weren't sure how we were going to do the code portions yet, so I left it as static content. I was going to add the `{.REPL}` and `{.snippet}` blocks after we had it figured out, so I can add them in now.

helmoski commented on content/efcore/getting-started/querying.md:
```diff
@@ -0,0 +1,47 @@
+
+## Loading All Data
+
+Grabs the entirety of a table.
```
Normally I am all for using as few words as possible, but this seems too terse. At this point, it is no different than the docs, in which case we should just be adding interactivity to the docs. Do you have any ideas for how to increase the content? To be clear, I don't want to add fluff; it should still be useful information.

helmoski replied:
Maybe describe the example. Full sentences would be better

helmoski commented on content/efcore/getting-started/querying.md:
```diff
@@ -0,0 +1,47 @@
+
+## Loading All Data
+
+Grabs the entirety of a table.
```
Where are we going to describe the database?

rdahlman3 replied:
We'll describe the database before all of the examples. I'll probably wait to add that in until I hear back from Maria as to whether or not we can use the database I've put together.

helmoski replied:
Ok, I'm going to wait to approve this PR until we have that content.

helmoski commented on content/efcore/getting-started/querying.md:
````diff
@@ -0,0 +1,47 @@
+```c#
+using (var context = new BlogsContext())
+{
+    var blogs = context.Blog.ToList();
````
`context.Blogs.ToList();` is the standard because you are working with multiple blogs rather than just one.

rdahlman3 replied:
I'll have to change that to Tracks anyway, so I'll fix that!

helmoski commented on content/efcore/getting-started/querying.md:
```diff
@@ -0,0 +1,47 @@
+using (var context = new BlogsContext())
+{
+    var blog = context.Blog
+        .Single(b => b.BlogId == 1); //grabs the Blog with a BlogId of 1
```
Usually queries are only put on multiple lines if there are multiple parts to the query. You can put the comment on the line above the query to avoid the line being too long.

rdahlman3 replied:
I believe the reason this was on a separate line was because that's how it was in the doc, so we assumed that was best practice.

The examples in the docs:

> using (var context = new BloggingContext())
> {
>    var blog = context.Blogs
>        .Single(b => b.BlogId == 1);
> }
> 
> using (var context = new BloggingContext())
> {
>    var blogs = context.Blogs
>        .Where(b => b.Url.Contains("dotnet"))
>        .ToList();
> }

helmoski replied:
Hmm... I don't know how I feel about that, but if that's how they do it...

helmoski commented on content/efcore/getting-started/toc.md:
```diff
@@ -0,0 +1 @@
+- [Querying the Database](querying.md)
```
This should match the title of of querying.md

rdahlman3 committed:
changed some things about the content. Still have to add code blocks to make it runnable.

rdahlman3 reviewed (commented):
Content has been edited - still need to add in runnable code blocks.

rdahlman3 committed:
fix some wording in the fetch all data section that I didn't like

arynhuck commented:
Reviewed. Everything looks good. I'm pulling in master.

arynhuck committed:
merge master -> 3-query_data_content

Brought branch up-to-date. Modified Topics Covered in README to include
link (also got rid of non-existent pages).

arynhuck committed:
merge master -> 3-query_data_content

This brings in changes from 16-what_is_ef_core

arynhuck commented:
Branch brought up-to-date with latest version of master. :shipit: ?

arynhuck committed:
merge master -> branch

Merge commit

arynhuck committed:
Content update

Updated content to be more explain-y and to include the database tables
the users will be able to query.

helmoski commented on content/efcore/getting-started/querying.md:
```diff
@@ -0,0 +1,91 @@
+ 
+Once we have an instance of the context, we can use it to interact with the database. To access the books in the database, we reference the relevant `DbSet` within the context - `Books` in our case - and call the `ToList` method to convert the `DbSet` to a `List`. The resulting list will contain all of the books within the database. 
+ 
+ 
```
Remove extra newline

helmoski commented on content/efcore/getting-started/querying.md:
```diff
@@ -0,0 +1,91 @@
+ 
+Some may note that we could retrieve all of the books like in the previous example and then filter them in our application. The problem with this is that it would require us to load all of the books into memory, and it also doesn't allow us to take advantage of our database's optimized querying functionality. Allowing the database to do what it does best and perform the filtering for us results i ...
+ 
+ 
```
Remove extra line

helmoski commented on content/efcore/getting-started/querying.md:
```diff
@@ -0,0 +1,91 @@
+# Querying Data 
+ 
+In this lesson, you'll learn how to use querying to fetch one or more items from your database. EF Core uses LINQ to query data, so it is important that you understand LINQ before going through this lesson. Check out the [LINQ lesson](../../csharp/getting-started/linq) in the C# Interactive Tutorial if you need a refresher. 
```
The link to the LINQ lesson is broken. I think it is just missing the file extension (linq.md)

davidmikh committed:
Removed extra lines & broken link to LINQ lesson

davidmikh commented:
I went over the latest reviews posted and updated querying.md accordingly.

helmoski commented:
We just came up with a design document for how multifile code examples will work. Please update this branch to follow the [design document](https://slack-files.com/T04MPFLR8-F2Y3TACB1-134d5e5bae). Then this PR should be good to go.

rdahlman3 commented:
Changes in the word doc need to be made.

- [ Change filtering section ]

rdahlman3 reviewed (commented):
Update filtering section to reflect comments on onedrive doc.

deakgeek committed:
Added samples folder, created readme and program files for code snippets: filtering entities, loading all entities, loading a single entity

helmoski commented on content/efcore/getting-started/samples/filtering-entities/Program.cs:
```diff
@@ -0,0 +1,21 @@
+
+            foreach (Book book in books)
+            {
+                Console.WriteLine(book.Genre);
```
Could we print out the book title here? 
or maybe both the title and genre? 
Otherwise, we will just be printing "Science Fiction" several times.

deakgeek replied:
Resolved

helmoski commented on content/efcore/getting-started/samples/filtering-entities/README.md:
```diff
@@ -0,0 +1,9 @@
+# Filtering Entities
+
+This code example demonstrates how to use EF Core to query a subset of entities based off filters from the database.
```
"based off filters from the database" sounds like we are getting filters from the database. 

How about "This code example demonstrates how to use EF Core to query a subset of entities from the database based on specified filters."

deakgeek replied:
Resolved

helmoski commented on content/efcore/getting-started/samples/loading-all-entities/README.md:
```diff
@@ -0,0 +1,10 @@
+- [LibraryContext.cs](../shared/LibraryContext.cs)
+- [Books.cs](../shared/Books.cs)
+- [Author.cs](../shared/Author.cs)
+
```
Could you remove the extra newline please?

deakgeek replied:
Resolved

helmoski commented on content/efcore/getting-started/samples/loading-single-entities/Program.cs:
```diff
@@ -0,0 +1,17 @@
+using System;
```
Could you rename the directory to `loading-single-entity`

deakgeek replied:
Resolved

helmoski commented on content/efcore/getting-started/samples/loading-single-entities/Program.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        using (var context = new LibraryContext())
+        {
+            //grabs the Title with a BookId of 1
```
It grabs the Book, not the Title. It prints the title afterwards.

I don't think this comment is necessary, though. The code is self-documenting.

deakgeek replied:
Resolved

deakgeek committed:
Updated filtering section on querying tutorial

helmoski commented on content/efcore/getting-started/samples/shared/Book.cs:
```diff
@@ -0,0 +1,8 @@
+public class Book
+{
+    public int BookId { get; set; }
+    public int AuthorId { get; set; }
```
Remove `AuthorId` field and add the following field to the end of this file:

``` C#
public Author Author { get; set; }
```

deakgeek replied:
Resolved

deakgeek committed:
Updated querying code snippets with requested changes

helmoski commented on content/efcore/getting-started/samples/shared/Book.cs:
```diff
@@ -1,8 +1,9 @@
\ No newline at end of file
+}
+
+public Author Author { get; set; }
```
Sorry, this is what I meant:

``` C#
public class Book
{
    public int BookId { get; set; }
    public string Title { get; set; }
    public string Genre { get; set; }
    public int PublicationYear { get; set; }

    public Author Author { get; set; }
}
```

helmoski commented on content/efcore/getting-started/samples/filtering-entities/Program.cs:
```diff
@@ -0,0 +1,21 @@
+
+            foreach (Book book in books)
+            {
+                Console.WriteLine(book.Title + ' ' + book.Genre);
```
Please use `String.format()`

helmoski commented on content/efcore/getting-started/samples/filtering-entities/README.md:
```diff
@@ -0,0 +1,9 @@
+## Code Files:
+- [Program.cs](Program.cs)
+- [LibraryContext.cs](../shared/LibraryContext.cs)
+- [Books.cs](../shared/Books.cs)
```
Book, not Books

helmoski commented on content/efcore/getting-started/samples/shared/Book.cs:
```diff
@@ -0,0 +1,9 @@
+    public int PublicationYear { get; set; }
+}
+
+public Author Author { get; set; }
```
Sorry, this is what I meant:

```C#
public class Book
{
    public int BookId { get; set; }
    public string Title { get; set; }
    public string Genre { get; set; }
    public int PublicationYear { get; set; }

    public Author Author { get; set; }
}
```

deakgeek committed:
Updated readme to correct path, updated other requested changes

helmoski commented on content/efcore/getting-started/samples/shared/LibraryContext.cs:
```diff
@@ -0,0 +1,13 @@
+    public DbSet<Book> Books { get; set; }
+    public DbSet<Author> Authors { get; set; }
+
+    protected override vdeaoid OnConfiguring(DbContextOptionsBuilder optionsBuilder)
```
typo, should be `void`

helmoski commented on content/efcore/getting-started/samples/filtering-entities/Program.cs:
```diff
@@ -0,0 +1,21 @@
+
+public class Program
+{
+    public static void Main(string[] args)
```
Remove `string[] args`. It's breaking the program.

helmoski commented on content/efcore/getting-started/samples/loading-all-entities/Program.cs:
```diff
@@ -0,0 +1,18 @@
+
+public class Program
+{
+    public static void Main(string[] args)
```
Remove `string[] args`. It's breaking the program.

helmoski commented on content/efcore/getting-started/samples/loading-single-entity/Program.cs:
```diff
@@ -0,0 +1,16 @@
+
+public class Program
+{
+    public static void Main(string[] args)
```
Remove `string[] args`. It's breaking the program.

rdahlman3 committed:
Remove string args and fix void typo

helmoski commented:
:shipit:
